File: 207-course-schedule/course-schedule.rb
# @param {Integer} num_courses
# @param {Integer[][]} prerequisites
# @return {Boolean}
def can_finish(num_courses, prerequisites)
  graph = Hash.new { |h, k| h[k] = [] }
  indegrees = [0] * num_courses
  prerequisites.each do |x, y|
    graph[x] << y
    indegrees[y] += 1
  end
  queue = []
  order = []
  print(indegrees, "\n")
  indegrees.each_with_index { |el, index| queue << index if el == 0 }
  print(queue, "\n")
  while !queue.empty?
    el = queue.shift
    order  << el
    graph[el].each do |node|
      if indegrees[node] != 0
        indegrees[node] -= 1
        queue << node if indegrees[node] == 0
      end
    end
  end
  order.size == num_courses
end

File: 207-course-schedule/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        visited = [0] * numCourses
        graph = defaultdict(list)
        def dfs(i):
            visited[i] = 1
            for child in graph[i]:
                if visited[child] == 1:
                    return True
                if visited[child] == 0 and dfs(child):
                    return True
            visited[i] = 2
            return False
        
        for u, v in prerequisites:
            graph[v].append(u)

        for i in range(numCourses):
            if visited[i] == 0:
                has_cycle = dfs(i)
                if has_cycle:
                    return False
        return True

        